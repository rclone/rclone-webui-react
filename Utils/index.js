import checkPropTypes from 'check-prop-types';
import {applyMiddleware, createStore} from 'redux';
import rootReducer from './../src/reducers';
import {middleware} from '../src/store';

export const findByTestAttr = (component, attr) => {
    const wrapper = component.find(`[data-test='${attr}']`);
    return wrapper;
};

export const checkProps = (component, expectedProps) => {
    const propsErr = checkPropTypes(component.propTypes, expectedProps, 'prop', component.name);
    return propsErr;
};

export const testStore = (initialState) => {
    const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
    return createStoreWithMiddleware(rootReducer, initialState);
};