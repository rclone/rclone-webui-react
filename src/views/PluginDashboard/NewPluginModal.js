import React, {useState} from 'react';
import {
    Button,
    Col,
    Form,
    FormFeedback,
    FormGroup,
    Input,
    Label,
    Modal,
    ModalBody,
    ModalFooter,
    ModalHeader
} from 'reactstrap';
import * as PropTypes from "prop-types"

/**
 * NewPluginModal shows a button for opening a modal for new mount and then executes okHandle when positive
 * button is clicked
 * @param props
 * @returns {*}
 * @constructor
 */
const NewPluginModal = (props) => {
    const {
        buttonLabel,
        className,
        okHandle
    } = props;

    const [modal, setModal] = useState(false);

    const [pluginDownloadURL, setPluginDownloadURL] = useState("");

    const [pluginType, setPluginType] = useState("");

    const toggle = () => setModal(!modal);

    const handleCreateMount = () => {
        if (!okHandle) {
            throw new Error("Ok handle is null");
        }
        okHandle(pluginDownloadURL, pluginType);
    }

    const isCreateDisabled = () => {
        return !pluginDownloadURL || pluginDownloadURL === "";
    }

    const submitForm = (e) => {
        e.preventDefault();
        handleCreateMount();
    }

    return (
        <div data-test="newMountModalComponent">
            <Button color="primary" onClick={toggle}>{buttonLabel}</Button>
            <Form onSubmit={submitForm}>
                <Modal isOpen={modal} toggle={toggle} className={className}>
                    <ModalHeader toggle={toggle}>New Mount</ModalHeader>
                    <ModalBody>

                        <FormGroup row>
                            <Label for={"pluginURL"} sm={5}>Plugin URL</Label>
                            <Col sm={7}>
                                <Input type={"text"} value={pluginDownloadURL}
                                       name={"pluginURL"}
                                       id={"pluginURL"} onChange={e => setPluginDownloadURL(e.target.value)}
                                       required={true}
                                >
                                </Input>
                                <FormFeedback/>
                            </Col>
                        </FormGroup>

                        <FormGroup row>
                            <Label for={"pluginType"} sm={5}>Plugin Type</Label>
                            <Col sm={7}>
                                <Input type={"text"} value={pluginDownloadURL}
                                       name={"pluginType"}
                                       id={"pluginType"} onChange={e => setPluginType(e.target.value)}
                                       required={true}>
                                </Input>
                                <FormFeedback/>
                            </Col>
                        </FormGroup>
                    </ModalBody>
                    <ModalFooter>
                        <Button data-test="ok-button" color="primary" onClick={handleCreateMount}
                                disabled={isCreateDisabled()}>Add</Button>{' '}
                        <Button data-test="cancel-button" color="secondary" onClick={toggle}>Cancel</Button>
                    </ModalFooter>
                </Modal>
            </Form>
        </div>
    );
}

NewPluginModal.propTypes = {
    /**
     * Text for open modal button
     */
    buttonLabel: PropTypes.string,
    /**
     * Class for open modal button
     */
    buttonClass: PropTypes.string,
    /**
     * Function to be called when ok button is clicked.
     */
    okHandle: PropTypes.func.isRequired,
}

export default NewPluginModal;