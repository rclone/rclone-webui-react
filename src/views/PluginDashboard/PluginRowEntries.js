import React, {useState} from "react";
import * as PropTypes from "prop-types";
import {Button} from "reactstrap";
import axiosInstance from "../../utils/API/API";
import {toast} from "react-toastify";

function PluginRowEntries({loadedPlugins, getPlugins}) {
	const [isLoading, setIsLoading] = useState(false);
	const removePlugin = (key) => {
		setIsLoading(true);
		axiosInstance.post("pluginsctl/removePlugin", {
			name: key
		}).then(res => {
			getPlugins();

			setIsLoading(false);
			toast.info(`Plugin ${key} removed successfully`);
		}, err => {
			setIsLoading(false);

			toast.error(`Error removing plugin ${key} : ${err}`);
		})

	}
	let entries = [];
	for (const [key, value] of Object.entries(loadedPlugins)) {
		entries.push(
			<tr key={key}>
				<td>{value.name}</td>
				<td>{value.author}</td>
				<td>{value.description}</td>
				<td>
					{isLoading ? "Installing..." :
						<>
							<Button color={"danger"} className="ml-2" onClick={() => removePlugin(key)}><span
								className="fa fa-trash pr-1"/>Delete</Button>
						</>
					}
				</td>
			</tr>
		)
	}
	return entries;
}

PluginRowEntries.propTypes = {
	loadedPlugins: PropTypes.object.isRequired,
	getPlugins: PropTypes.func.isRequired
}

export default PluginRowEntries;
