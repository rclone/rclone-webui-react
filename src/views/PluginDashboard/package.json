{
  "name": "PluginDashboard",
  "version": "0.0.0",
  "private": true,
  "main": "PluginDashboard.js"
}
