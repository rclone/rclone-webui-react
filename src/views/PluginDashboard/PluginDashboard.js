import React from 'react';
import {connect} from "react-redux";
import {Button, Col, Row, Table} from "reactstrap";
import * as PropTypes from 'prop-types';
import {addPlugin, getPlugins} from "../../actions/pluginActions";
import PluginRowEntries from "./PluginRowEntries";
import NewPluginModal from "./NewPluginModal";

class PluginDashboard extends React.Component {

	componentDidMount() {
		const {getPlugins} = this.props;
		getPlugins();
	}

	addPluginHandle = (pluginDownloadURL, pluginType) => {
		const {addPlugin} = this.props;
		addPlugin(pluginDownloadURL, pluginType);
	}

	render() {
		const {loadedPlugins, getPlugins} = this.props;
		return (
			<div data-test="pluginDashboardComponent">
				<Row>
					<Col lg={12} className="mb-4 d-flex justify-content-between">
						<NewPluginModal buttonLabel="Add New" okHandle={this.addPluginHandle}/>
						<Button color="primary" onClick={() => this.props.history.push("/storeDashboard")}>
							Visit Store
						</Button>
					</Col>
				</Row>
				<Table responsive className="table-striped">
					<thead>
					<tr>
						<th>Name</th>
						<th>Author</th>
						<th>Description</th>
						<th>Actions</th>
					</tr>
					</thead>
					<tbody>
					<PluginRowEntries loadedPlugins={loadedPlugins} getPlugins={getPlugins}/>
					</tbody>
				</Table>
			</div>);
	}
}

const mapStateToProps = state => ({
	loadedPlugins: state.plugins.loadedPlugins,
});

PluginDashboard.propTypes = {
	loadedPlugins: PropTypes.object.isRequired,

	getPlugins: PropTypes.func.isRequired,

	addPlugin: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, {getPlugins, addPlugin})(PluginDashboard);
