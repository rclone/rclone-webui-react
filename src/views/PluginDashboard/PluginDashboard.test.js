import React from "react";
import {shallow} from "enzyme";
import {findByTestAttr, testStore} from "../../../Utils";
import PluginDashboard from "./PluginDashboard";
import toJson from "enzyme-to-json";


const setUp = (intialState = {}, props = {}) => {
	const store = testStore(intialState);

	const component = shallow(<PluginDashboard {...props} store={store}/>);
	return component.childAt(0).dive();
};


describe('Plugin Dashboard', function () {

	describe('renders', function () {
		let wrapper;
		beforeEach(() => {
			const initialState = {
				status: {
					checkStatus: true
				}
			};

			const props = {};
			wrapper = setUp(initialState, props)
		});

		it('should render without crashing', function () {
			const component = findByTestAttr(wrapper, "pluginDashboardComponent");
			expect(component).toHaveLength(1);
		});

		it('should match snapshot', function () {
			expect(toJson(wrapper)).toMatchSnapshot()
		});
	});

});
