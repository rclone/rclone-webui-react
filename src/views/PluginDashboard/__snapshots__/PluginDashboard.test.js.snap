// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin Dashboard renders should match snapshot 1`] = `
<div
  data-test="pluginDashboardComponent"
>
  <Row
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Col
      className="mb-4 d-flex justify-content-between"
      lg={12}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <NewPluginModal
        buttonLabel="Add New"
        okHandle={[Function]}
      />
      <Button
        color="primary"
        onClick={[Function]}
        tag="button"
      >
        Visit Store
      </Button>
    </Col>
  </Row>
  <Table
    className="table-striped"
    responsive={true}
    responsiveTag="div"
    tag="table"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Author
        </th>
        <th>
          Description
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <PluginRowEntries
        getPlugins={[Function]}
        loadedPlugins={Object {}}
      />
    </tbody>
  </Table>
</div>
`;
